File: 668 Kth Smallest Number in Multiplication Table.c


int findKthNumber(int m, int n, int k){
        int start = 1;
        int end = m * n;
        while (start < end){
            int mid = round((start + end)/2);
            int curr = 0;
            int j = 1;
            for (int r=m; r>0 ;r--){
                while (j <= n&& r * j <= mid){
                    j += 1;
                }
                curr += j-1;
            }
            if (curr < k){
                start = mid + 1;
            } else{
                end = mid;
            }
        }   
        return start;
        
}

File: 668 Kth Smallest Number in Multiplication Table.java
class Solution {
    public int findKthNumber(int m, int n, int k) {
        int start = 1, end = m * n + 1;
        while (start < end) {
            int mid = start + (end - start) / 2;
            if (!check(m, n, mid, k)){
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return start;
    }
    
    private boolean check(int m, int n, int v, int k) {
        int num = 0;
        for (int i = 1, j = n; i <= m; ++i) {
            while (j >= 1 && j * i > v) {
                j--;
            }               
            if (j == 0) {
                break;
            }
            num += j;
            if (num >= k) {
                return true;
            }
        }
        return false;
    }
    
}
